src/robot/GradientDescentIk.cpp

```diff
@@ -0,0 +1,61 @@
+#include "aikido/robot/GradientDescentIk.hpp"
+
+#include "aikido/constraint/dart/JointStateSpaceHelpers.hpp"
+#include "aikido/statespace/dart/MetaSkeletonStateSaver.hpp"
+
+using aikido::constraint::SampleGenerator;
+using aikido::constraint::dart::createSampleableBounds;
+using aikido::statespace::dart::MetaSkeletonStateSaver;
+
+namespace aikido {
+namespace robot {
+
+//==============================================================================
+GradientDescentIk::GradientDescentIk(
+    ::dart::dynamics::MetaSkeletonPtr arm,
+    ::dart::dynamics::BodyNodePtr endEffector,
+    statespace::dart::ConstMetaSkeletonStateSpacePtr armSpace,
+    common::RNG* rng)
+  : mArm(arm)
+  , mEndEffector(endEffector)
+  , mArmSpace(armSpace)
+  , mSeedConfigSampleable(
+        createSampleableBounds(mArmSpace, std::move(cloneRNGFrom(*rng)[0])))
+  , mDartIkSolver(::dart::dynamics::InverseKinematics::create(mEndEffector))
+{
+  // Do nothing.
+}
+
+//==============================================================================
+std::vector<Eigen::VectorXd> GradientDescentIk::getSolutions(
+    const Eigen::Isometry3d& targetPose, const size_t maxSolutions) const
+{
+  auto saver = MetaSkeletonStateSaver(
+      mArm, MetaSkeletonStateSaver::Options::POSITIONS);
+  DART_UNUSED(saver);
+
+  std::vector<Eigen::VectorXd> solutions;
+
+  std::shared_ptr<SampleGenerator> ikSeedGenerator
+      = mSeedConfigSampleable->createSampleGenerator();
+  auto seedState = mArmSpace->createState();
+
+  for (size_t i = 0; i < maxSolutions; i++)
+  {
+    // TODO: What to do if we can't sample a seed config?
+    if (!ikSeedGenerator->sample(seedState))
+      continue;
+
+    mArmSpace->setState(mArm.get(), seedState);
```

Comment:
Just to verify that I'm understanding this right, the only functional difference between this GradientDescentSolver and DART's GradientDescentSolver is that this one samples a random state as the initial guess, right?

Reply:
In the code [here](https://github.com/dartsim/dart/blob/7e3db63a07f7fcabd577efbd0487c6cd99e31c34/dart/dynamics/InverseKinematics.cpp#L103) DART set's the problem's initial guess to the current state, and then in the code [here](https://github.com/dartsim/dart/blob/7e3db63a07f7fcabd577efbd0487c6cd99e31c34/dart/optimizer/GradientDescentSolver.cpp#L132) it sets that value to the initial guess

Reply:
In our current stack, [randomizing the state is done in AIKIDO's InverseKinematicsSampleable](https://github.com/personalrobotics/aikido/blob/c0a8b7eff286640374631a2c6bddb571b829a73e/src/constraint/dart/InverseKinematicsSampleable.cpp#L209), which calls DART's IK class. So we would have to change InverseKinematicsSampleable accordingly.

(Regardless, if we intend AIKIDO's IK class to become our main IK class as opposed to the DART one, we should change InverseKinematicsSampleable to work with AIKIDO's IK class by default. InverseKinematicsSampleable is effectively the class that takes a pose-solving IK and generalizes it to a TSR-solving IK.)

Reply:
Also, another question is should `InverseKinematicsSampleable` be in the same folder/namespace as the rest of the IK?

Reply:
Yeah! This is really just wrapping DART's IK solver in the new interface and randomizing the initial state each time (again since DART just goes from the current position of the skeleton).

And yeah, we will need to rewrite `InverseKinematicsSampleable`- my goal is to merge this new API in this PR, and then start actually using it in subsequent ones. As for where to put that class I think it belongs where it is with the other sampleable stuff- IK is more of a robot thing IMHO